harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -601,17 +601,31 @@ def make_workloads() -> workload.Stream:
         )
         controller.run()
 
-    @pytest.mark.skipif(not HAVE_APEX or not torch.cuda.is_available(), reason="no gpu available")
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="no gpu available")
     @pytest.mark.gpu
-    def test_apex_amp(self) -> None:
-        apex.amp.register_float_function(torch, "sigmoid")
+    @pytest.mark.parametrize(
+        "trial_class",
+        [
+            pytorch_onevar_model.OneVarApexAMPTrial,
+            pytorch_onevar_model.OneVarAutoAMPTrial,
+            pytorch_onevar_model.OneVarManualAMPTrial,
+        ],
+        ids=[
+            "apex",
+            "autocast",
+            "manual",
+        ],
+    )
+    def test_amp(self, trial_class) -> None:
```

Comment:
do we assert / somewhere that the losses are what we expect after AMP/scaling? am i missing something?

Reply:
You're not missing anything. The code is. Stay tuned!

Reply:
I think there's still some assertions missing here.

The assert_output_float16 assertion seems good.

All the math is written in the top of pytorch_onevar_model.py to calculate what the exact expected gradient updates would be.  An example of what I have in mind is how `TestPyTorchTrial::test_onevar_single()` tests each gradient step.

Another good example is `e2e/tests/experiment/test_pytorch_distributed.py::test_gradient_aggregation()`, which also calculates all of the expected gradient updates and makes sure the actual updates are within computational error of the expected ones.

Reply:
This is done.